File: pipeline/reimpl/coactvwc.py
"""
Reimplementation of COACTVWC — CardDemo Account View Screen.

CICS online program. Accepts an account ID, reads account and customer
data, and displays the full account detail view.

Navigation:
  ENTER → look up account by ID and display
  PF3   → return to calling program / main menu
"""

from __future__ import annotations
from dataclasses import dataclass, field
from typing import Optional

from .carddemo_data import (
    CarddemoCommarea, ScreenHeader, AccountRecord, CustomerRecord, CardXrefRecord,
    CCDA_MSG_INVALID_KEY, DFHENTER, DFHPF3,
)


WS_PGMNAME   = "COACTVWC"
WS_TRANID    = "CAVW"
LIT_MENUPGM  = "COMEN01C"


# ── Screen output ─────────────────────────────────────────────────────────────

@dataclass
class AccountViewData:
    """Screen fields for account detail display."""
    acct_id: int = 0
    acct_active_status: str = ""
    acct_curr_bal: str = ""
    acct_credit_limit: str = ""
    acct_cash_credit_limit: str = ""
    acct_curr_cyc_credit: str = ""
    acct_curr_cyc_debit: str = ""
    acct_open_date: str = ""
    acct_expiration_date: str = ""
    acct_reissue_date: str = ""
    acct_group_id: str = ""
    cust_id: int = 0
    cust_ssn_masked: str = ""
    cust_name: str = ""
    cust_dob: str = ""
    cust_fico_score: int = 0
    cust_addr_zip: str = ""
    cust_phone_1: str = ""
    cust_phone_2: str = ""


@dataclass
class AccountViewResult:
    account_data: Optional[AccountViewData] = None
    commarea: Optional[CarddemoCommarea] = None
    screen: Optional[ScreenHeader] = None
    message: str = ""
    error: bool = False
    xctl_program: str = ""
    return_to_prev: bool = False


# ── Repository interfaces ──────────────────────────────────────────────────────

class AccountRepository:
    def __init__(self, accounts: dict[int, AccountRecord]):
        self._accounts = accounts

    def find(self, acct_id: int) -> Optional[AccountRecord]:
        return self._accounts.get(acct_id)


class XrefRepository:
    def __init__(self, xrefs_by_acct: dict[int, CardXrefRecord]):
        self._by_acct = xrefs_by_acct

    def find_by_acct(self, acct_id: int) -> Optional[CardXrefRecord]:
        return self._by_acct.get(acct_id)


class CustomerRepository:
    def __init__(self, customers: dict[int, CustomerRecord]):
        self._customers = customers

    def find(self, cust_id: int) -> Optional[CustomerRecord]:
        return self._customers.get(cust_id)


# ── Core logic ────────────────────────────────────────────────────────────────

def process_account_view(
    eibcalen: int,
    eibaid: str,
    commarea: CarddemoCommarea,
    acct_id_input: str,
    account_repo: AccountRepository,
    xref_repo: XrefRepository,
    customer_repo: CustomerRepository,
) -> AccountViewResult:
    """Process account view screen — mirrors COACTVWC PROCEDURE DIVISION."""
    result = AccountViewResult()

    # Always set up commarea context
    if eibcalen == 0 or (
        commarea.cdemo_from_program == LIT_MENUPGM
        and commarea.cdemo_pgm_context == 0
    ):
        commarea = CarddemoCommarea()

    result.commarea = commarea

    if eibaid in (DFHPF3, "PF3"):
        # Return to calling program
        back_pgm = commarea.cdemo_from_program or LIT_MENUPGM
        back_tranid = commarea.cdemo_from_tranid or "CM00"
        commarea.cdemo_to_program = back_pgm
        commarea.cdemo_to_tranid = back_tranid
        commarea.cdemo_from_tranid = WS_TRANID
        commarea.cdemo_from_program = WS_PGMNAME
        commarea.cdemo_pgm_context = 0
        result.xctl_program = back_pgm
        result.return_to_prev = True
        result.commarea = commarea
        return result

    # ENTER or any other key → process input and display
    if commarea.cdemo_pgm_context == 0:
        # First entry — show blank search form
        commarea.cdemo_pgm_context = 1
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.message = "Enter or update id of account to display"
        result.screen.errmsg = result.message
        result.commarea = commarea
        return result

    # Re-entry with account ID
    return _process_account_lookup(
        acct_id_input, account_repo, xref_repo, customer_repo, commarea, result
    )


def _process_account_lookup(
    acct_id_input: str,
    account_repo: AccountRepository,
    xref_repo: XrefRepository,
    customer_repo: CustomerRepository,
    commarea: CarddemoCommarea,
    result: AccountViewResult,
) -> AccountViewResult:
    """2000-PROCESS-INPUTS + 9000-READ-ACCT."""
    # Validate input
    if not acct_id_input or acct_id_input.strip() == "":
        result.error = True
        result.message = "Account number not provided"
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.screen.errmsg = result.message
        result.commarea = commarea
        return result

    try:
        acct_id = int(acct_id_input.strip())
    except ValueError:
        result.error = True
        result.message = "Account number must be a non zero 11 digit number"
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.screen.errmsg = result.message
        result.commarea = commarea
        return result

    if acct_id == 0:
        result.error = True
        result.message = "Account number must be a non zero 11 digit number"
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.screen.errmsg = result.message
        result.commarea = commarea
        return result

    # Look up account
    acct = account_repo.find(acct_id)
    if acct is None:
        result.error = True
        result.message = "Did not find this account in account master file"
        result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
        result.screen.errmsg = result.message
        result.commarea = commarea
        return result

    # Look up customer via xref
    xref = xref_repo.find_by_acct(acct_id)
    cust = None
    if xref:
        cust = customer_repo.find(xref.xref_cust_id)
    if cust is None:
        result.error = True
        result.message = "Did not find associated customer in master file"

    view_data = _build_view_data(acct, cust)
    result.account_data = view_data
    result.screen = ScreenHeader.now(WS_TRANID, WS_PGMNAME)
    result.screen.errmsg = result.message if result.error else "Displaying details of given Account"
    result.commarea = commarea
    return result


def _build_view_data(acct: AccountRecord, cust: Optional[CustomerRecord]) -> AccountViewData:
    vd = AccountViewData(
        acct_id=acct.acct_id,
        acct_active_status=acct.acct_active_status,
        acct_curr_bal=f"{float(acct.acct_curr_bal):,.2f}",
        acct_credit_limit=f"{float(acct.acct_credit_limit):,.2f}",
        acct_cash_credit_limit=f"{float(acct.acct_cash_credit_limit):,.2f}",
        acct_curr_cyc_credit=f"{float(acct.acct_curr_cyc_credit):,.2f}",
        acct_curr_cyc_debit=f"{float(acct.acct_curr_cyc_debit):,.2f}",
        acct_open_date=acct.acct_open_date,
        acct_expiration_date=acct.acct_expiration_date,
        acct_reissue_date=acct.acct_reissue_date,
        acct_group_id=acct.acct_group_id,
    )
    if cust:
        ssn = str(cust.cust_ssn).zfill(9)
        vd.cust_id = cust.cust_id
        vd.cust_ssn_masked = f"{ssn[:3]}-{ssn[3:5]}-{ssn[5:]}"
        vd.cust_name = f"{cust.cust_first_name.strip()} {cust.cust_last_name.strip()}"
        vd.cust_dob = cust.cust_dob_yyyy_mm_dd
        vd.cust_fico_score = cust.cust_fico_credit_score
        vd.cust_addr_zip = cust.cust_addr_zip
        vd.cust_phone_1 = cust.cust_phone_num_1
        vd.cust_phone_2 = cust.cust_phone_num_2
    return vd


# ── Differential harness runner adapter ──────────────────────────────────────

from decimal import Decimal as _Decimal

_SEED_ACCOUNTS: dict[int, AccountRecord] = {
    10000001: AccountRecord(
        acct_id=10000001, acct_active_status="Y",
        acct_curr_bal=_Decimal("1500.00"), acct_credit_limit=_Decimal("5000.00"),
        acct_cash_credit_limit=_Decimal("1000.00"),
        acct_open_date="2020-01-15", acct_expiration_date="2028-01-15",
        acct_reissue_date="2025-01-15",
        acct_curr_cyc_credit=_Decimal("200.00"), acct_curr_cyc_debit=_Decimal("50.00"),
        acct_group_id="GOLD",
    ),
}

_SEED_XREFS_BY_ACCT: dict[int, CardXrefRecord] = {
    10000001: CardXrefRecord(xref_card_num="4111111111110001", xref_cust_id=100001, xref_acct_id=10000001),
}

_SEED_CUSTOMERS: dict[int, CustomerRecord] = {
    100001: CustomerRecord(
        cust_id=100001, cust_first_name="Jane", cust_last_name="User",
        cust_ssn=123456789, cust_dob_yyyy_mm_dd="1985-03-15",
        cust_fico_credit_score=750, cust_addr_zip="10001",
        cust_phone_num_1="212-555-0100", cust_phone_num_2="212-555-0101",
    ),
}


def run_vector(inputs: dict) -> dict:
    """Canonical runner entry point for the differential harness.

    SCENARIO selects a hardcoded test path:
      FIRST_ENTRY    — context=0, first entry → show blank form
      VALID_INPUT    — ENTER with valid account ID → display account
      ACCT_NOT_FOUND — ENTER with bad account ID → error
      PF3_RETURN     — press PF3 → return to menu
    """
    scenario = inputs.get("SCENARIO", "FIRST_ENTRY")

    account_repo = AccountRepository(dict(_SEED_ACCOUNTS))
    xref_repo = XrefRepository(dict(_SEED_XREFS_BY_ACCT))
    customer_repo = CustomerRepository(dict(_SEED_CUSTOMERS))

    commarea = CarddemoCommarea(
        cdemo_from_tranid="CM00",
        cdemo_from_program="COMEN01C",
        cdemo_user_id="USER0001",
        cdemo_user_type="U",
        cdemo_pgm_context=1,
    )

    if scenario == "FIRST_ENTRY":
        commarea.cdemo_pgm_context = 0
        result = process_account_view(
            eibcalen=100, eibaid="ENTER", commarea=commarea,
            acct_id_input="",
            account_repo=account_repo, xref_repo=xref_repo,
            customer_repo=customer_repo,
        )
    elif scenario == "VALID_INPUT":
        result = process_account_view(
            eibcalen=100, eibaid="ENTER", commarea=commarea,
            acct_id_input="10000001",
            account_repo=account_repo, xref_repo=xref_repo,
            customer_repo=customer_repo,
        )
    elif scenario == "ACCT_NOT_FOUND":
        result = process_account_view(
            eibcalen=100, eibaid="ENTER", commarea=commarea,
            acct_id_input="99999999",
            account_repo=account_repo, xref_repo=xref_repo,
            customer_repo=customer_repo,
        )
    elif scenario == "PF3_RETURN":
        result = process_account_view(
            eibcalen=100, eibaid="PF3", commarea=commarea,
            acct_id_input="",
            account_repo=account_repo, xref_repo=xref_repo,
            customer_repo=customer_repo,
        )
    else:
        commarea.cdemo_pgm_context = 0
        result = process_account_view(
            eibcalen=100, eibaid="ENTER", commarea=commarea,
            acct_id_input="",
            account_repo=account_repo, xref_repo=xref_repo,
            customer_repo=customer_repo,
        )

    acct_id_out = ""
    acct_status_out = ""
    acct_bal_out = ""
    cust_name_out = ""
    cust_ssn_out = ""
    if result.account_data:
        acct_id_out = str(result.account_data.acct_id)
        acct_status_out = result.account_data.acct_active_status
        acct_bal_out = result.account_data.acct_curr_bal
        cust_name_out = result.account_data.cust_name
        cust_ssn_out = result.account_data.cust_ssn_masked

    return {
        "ERROR": "Y" if result.error else "N",
        "MESSAGE": result.message,
        "XCTL_PROGRAM": result.xctl_program or "",
        "RETURN_TO_PREV": "Y" if result.return_to_prev else "N",
        "ACCT_ID": acct_id_out,
        "ACCT_STATUS": acct_status_out,
        "ACCT_BAL": acct_bal_out,
        "CUST_NAME": cust_name_out,
        "CUST_SSN": cust_ssn_out,
    }

File: pipeline/reimpl/java/runner/src/main/java/com/modernization/masquerade/runner/programs/Coactvwc.java
package com.modernization.masquerade.runner.programs;

import com.modernization.masquerade.runner.ProgramRunner;

import java.math.BigDecimal;
import java.util.*;

/**
 * Java reimplementation of COACTVWC — CardDemo Account View Screen.
 *
 * <p>Mirrors {@code pipeline/reimpl/coactvwc.py}. Python is the source of truth.
 *
 * <p>Structure: standalone {@link #processAccountView} method + thin
 * {@link #runVector} adapter.
 */
public class Coactvwc implements ProgramRunner {

    // ── Inner data types ────────────────────────────────────────────────────

    static class AccountRecord {
        int acctId;
        String acctActiveStatus;
        BigDecimal acctCurrBal;
        BigDecimal acctCreditLimit;
        BigDecimal acctCashCreditLimit;
        String acctOpenDate;
        String acctExpirationDate;
        String acctReissueDate;
        BigDecimal acctCurrCycCredit;
        BigDecimal acctCurrCycDebit;
        String acctGroupId;

        AccountRecord(int acctId, String acctActiveStatus, BigDecimal acctCurrBal,
                      BigDecimal acctCreditLimit, BigDecimal acctCashCreditLimit,
                      String acctOpenDate, String acctExpirationDate, String acctReissueDate,
                      BigDecimal acctCurrCycCredit, BigDecimal acctCurrCycDebit, String acctGroupId) {
            this.acctId = acctId;
            this.acctActiveStatus = acctActiveStatus;
            this.acctCurrBal = acctCurrBal;
            this.acctCreditLimit = acctCreditLimit;
            this.acctCashCreditLimit = acctCashCreditLimit;
            this.acctOpenDate = acctOpenDate != null ? acctOpenDate : "";
            this.acctExpirationDate = acctExpirationDate != null ? acctExpirationDate : "";
            this.acctReissueDate = acctReissueDate != null ? acctReissueDate : "";
            this.acctCurrCycCredit = acctCurrCycCredit;
            this.acctCurrCycDebit = acctCurrCycDebit;
            this.acctGroupId = acctGroupId != null ? acctGroupId : "";
        }
    }

    static class CustomerRecord {
        int custId;
        String custFirstName;
        String custLastName;
        int custSsn;
        String custDobYyyyMmDd;
        int custFicoCreditScore;
        String custAddrZip;
        String custPhoneNum1;
        String custPhoneNum2;

        CustomerRecord(int custId, String custFirstName, String custLastName,
                       int custSsn, String custDobYyyyMmDd, int custFicoCreditScore,
                       String custAddrZip, String custPhoneNum1, String custPhoneNum2) {
            this.custId = custId;
            this.custFirstName = custFirstName != null ? custFirstName : "";
            this.custLastName = custLastName != null ? custLastName : "";
            this.custSsn = custSsn;
            this.custDobYyyyMmDd = custDobYyyyMmDd != null ? custDobYyyyMmDd : "";
            this.custFicoCreditScore = custFicoCreditScore;
            this.custAddrZip = custAddrZip != null ? custAddrZip : "";
            this.custPhoneNum1 = custPhoneNum1 != null ? custPhoneNum1 : "";
            this.custPhoneNum2 = custPhoneNum2 != null ? custPhoneNum2 : "";
        }
    }

    static class CardXrefRecord {
        String xrefCardNum;
        int xrefCustId;
        int xrefAcctId;

        CardXrefRecord(String xrefCardNum, int xrefCustId, int xrefAcctId) {
            this.xrefCardNum = xrefCardNum;
            this.xrefCustId = xrefCustId;
            this.xrefAcctId = xrefAcctId;
        }
    }

    static class AccountViewData {
        int acctId;
        String acctActiveStatus;
        String acctCurrBal;
        String custName;
        String custSsnMasked;

        AccountViewData(int acctId, String acctActiveStatus, String acctCurrBal,
                        String custName, String custSsnMasked) {
            this.acctId = acctId;
            this.acctActiveStatus = acctActiveStatus;
            this.acctCurrBal = acctCurrBal;
            this.custName = custName;
            this.custSsnMasked = custSsnMasked;
        }
    }

    static class AccountViewResult {
        AccountViewData accountData = null;
        String message = "";
        boolean error = false;
        String xctlProgram = "";
        boolean returnToPrev = false;
    }

    // ── Repositories ────────────────────────────────────────────────────────

    static class AccountRepository {
        private final Map<Integer, AccountRecord> accounts;

        AccountRepository(Map<Integer, AccountRecord> accounts) {
            this.accounts = new HashMap<>(accounts);
        }

        AccountRecord find(int acctId) {
            return accounts.get(acctId);
        }
    }

    static class XrefRepository {
        private final Map<Integer, CardXrefRecord> byAcct;

        XrefRepository(Map<Integer, CardXrefRecord> byAcct) {
            this.byAcct = new HashMap<>(byAcct);
        }

        CardXrefRecord findByAcct(int acctId) {
            return byAcct.get(acctId);
        }
    }

    static class CustomerRepository {
        private final Map<Integer, CustomerRecord> customers;

        CustomerRepository(Map<Integer, CustomerRecord> customers) {
            this.customers = new HashMap<>(customers);
        }

        CustomerRecord find(int custId) {
            return customers.get(custId);
        }
    }

    // ── Standalone business method ──────────────────────────────────────────

    /**
     * Process account view screen — mirrors {@code process_account_view} in
     * {@code pipeline/reimpl/coactvwc.py}.
     */
    AccountViewResult processAccountView(
            int eibcalen, String eibaid, int pgmContext, String fromProgram,
            String acctIdInput,
            AccountRepository accountRepo, XrefRepository xrefRepo, CustomerRepository customerRepo) {

        AccountViewResult result = new AccountViewResult();

        if ("PF3".equals(eibaid)) {
            String backPgm = (fromProgram != null && !fromProgram.isEmpty()) ? fromProgram : "COMEN01C";
            result.xctlProgram = backPgm;
            result.returnToPrev = true;
            return result;
        }

        if (pgmContext == 0) {
            // First entry — show blank search form
            result.message = "Enter or update id of account to display";
            return result;
        }

        // Re-entry with account ID — process lookup
        return processAccountLookup(acctIdInput, accountRepo, xrefRepo, customerRepo, result);
    }

    private AccountViewResult processAccountLookup(
            String acctIdInput,
            AccountRepository accountRepo, XrefRepository xrefRepo,
            CustomerRepository customerRepo,
            AccountViewResult result) {

        if (acctIdInput == null || acctIdInput.trim().isEmpty()) {
            result.error = true;
            result.message = "Account number not provided";
            return result;
        }

        int acctId;
        try {
            acctId = Integer.parseInt(acctIdInput.trim());
        } catch (NumberFormatException e) {
            result.error = true;
            result.message = "Account number must be a non zero 11 digit number";
            return result;
        }

        if (acctId == 0) {
            result.error = true;
            result.message = "Account number must be a non zero 11 digit number";
            return result;
        }

        AccountRecord acct = accountRepo.find(acctId);
        if (acct == null) {
            result.error = true;
            result.message = "Did not find this account in account master file";
            return result;
        }

        CardXrefRecord xref = xrefRepo.findByAcct(acctId);
        CustomerRecord cust = null;
        if (xref != null) {
            cust = customerRepo.find(xref.xrefCustId);
        }
        if (cust == null) {
            result.error = true;
            result.message = "Did not find associated customer in master file";
        }

        // Build view data
        String custName = "";
        String custSsnMasked = "";
        if (cust != null) {
            custName = cust.custFirstName.trim() + " " + cust.custLastName.trim();
            String ssn = String.format("%09d", cust.custSsn);
            custSsnMasked = ssn.substring(0, 3) + "-" + ssn.substring(3, 5) + "-" + ssn.substring(5);
        }

        String balFormatted = String.format("%,.2f", acct.acctCurrBal.doubleValue());

        result.accountData = new AccountViewData(
                acct.acctId, acct.acctActiveStatus, balFormatted, custName, custSsnMasked);

        if (!result.error) {
            result.message = "Displaying details of given Account";
        }

        return result;
    }

    // ── runVector adapter ───────────────────────────────────────────────────

    @Override
    public Map<String, String> runVector(Map<String, String> inputs) {
        String scenario = inputs.getOrDefault("SCENARIO", "FIRST_ENTRY");

        // Seed data
        Map<Integer, AccountRecord> seedAccounts = new HashMap<>();
        seedAccounts.put(10000001, new AccountRecord(10000001, "Y",
                new BigDecimal("1500.00"), new BigDecimal("5000.00"), new BigDecimal("1000.00"),
                "2020-01-15", "2028-01-15", "2025-01-15",
                new BigDecimal("200.00"), new BigDecimal("50.00"), "GOLD"));
        AccountRepository accountRepo = new AccountRepository(seedAccounts);

        Map<Integer, CardXrefRecord> seedXrefs = new HashMap<>();
        seedXrefs.put(10000001, new CardXrefRecord("4111111111110001", 100001, 10000001));
        XrefRepository xrefRepo = new XrefRepository(seedXrefs);

        Map<Integer, CustomerRecord> seedCustomers = new HashMap<>();
        seedCustomers.put(100001, new CustomerRecord(100001, "Jane", "User",
                123456789, "1985-03-15", 750, "10001", "212-555-0100", "212-555-0101"));
        CustomerRepository customerRepo = new CustomerRepository(seedCustomers);

        AccountViewResult result;

        switch (scenario) {
            case "FIRST_ENTRY":
                result = processAccountView(100, "ENTER", 0, "COMEN01C",
                        "", accountRepo, xrefRepo, customerRepo);
                break;
            case "VALID_INPUT":
                result = processAccountView(100, "ENTER", 1, "COMEN01C",
                        "10000001", accountRepo, xrefRepo, customerRepo);
                break;
            case "ACCT_NOT_FOUND":
                result = processAccountView(100, "ENTER", 1, "COMEN01C",
                        "99999999", accountRepo, xrefRepo, customerRepo);
                break;
            case "PF3_RETURN":
                result = processAccountView(100, "PF3", 1, "COMEN01C",
                        "", accountRepo, xrefRepo, customerRepo);
                break;
            default:
                result = processAccountView(100, "ENTER", 0, "COMEN01C",
                        "", accountRepo, xrefRepo, customerRepo);
                break;
        }

        String acctId = "";
        String acctStatus = "";
        String acctBal = "";
        String custName = "";
        String custSsn = "";
        if (result.accountData != null) {
            acctId = String.valueOf(result.accountData.acctId);
            acctStatus = result.accountData.acctActiveStatus;
            acctBal = result.accountData.acctCurrBal;
            custName = result.accountData.custName;
            custSsn = result.accountData.custSsnMasked;
        }

        Map<String, String> out = new LinkedHashMap<>();
        out.put("ERROR", result.error ? "Y" : "N");
        out.put("MESSAGE", result.message);
        out.put("XCTL_PROGRAM", result.xctlProgram != null ? result.xctlProgram : "");
        out.put("RETURN_TO_PREV", result.returnToPrev ? "Y" : "N");
        out.put("ACCT_ID", acctId);
        out.put("ACCT_STATUS", acctStatus);
        out.put("ACCT_BAL", acctBal);
        out.put("CUST_NAME", custName);
        out.put("CUST_SSN", custSsn);
        return out;
    }
}
